flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TriggerTestHarness.java

```diff
@@ -0,0 +1,369 @@
+
+	private final HeapKeyedStateBackend<Integer> stateBackend;
+	private final TestInternalTimerService<Integer, W> internalTimerService;
+
+	public TriggerTestHarness(
+			Trigger<T, W> trigger,
+			TypeSerializer<W> windowSerializer) throws Exception {
+		this.trigger = trigger;
+		this.windowSerializer = windowSerializer;
+
+		// we only ever use one key, other tests make sure that windows work across different
+		// keys
+		DummyEnvironment dummyEnv = new DummyEnvironment("test", 1, 0);
+		MemoryStateBackend backend = new MemoryStateBackend();
+
+		@SuppressWarnings("unchecked")
+		HeapKeyedStateBackend<Integer> stateBackend = (HeapKeyedStateBackend<Integer>) backend.createKeyedStateBackend(dummyEnv,
+				new JobID(),
+				"test_op",
+				IntSerializer.INSTANCE,
+				1,
+				new KeyGroupRange(0, 0),
+				new KvStateRegistry().createTaskRegistry(new JobID(), new JobVertexID()));
+		this.stateBackend = stateBackend;
+
+		this.stateBackend.setCurrentKey(0);
+
+		this.internalTimerService = new TestInternalTimerService<>(new KeyContext() {
+			@Override
+			public void setCurrentKey(Object key) {
+				// ignore
+			}
+
+			@Override
+			public Object getCurrentKey() {
+				return KEY;
+			}
+		});
+	}
+
+	public int numProcessingTimeTimers() {
+		return internalTimerService.numProcessingTimeTimers();
+	}
+
+	public int numProcessingTimeTimers(W window) {
+		return internalTimerService.numProcessingTimeTimers(window);
+	}
+
+	public int numEventTimeTimers() {
+		return internalTimerService.numEventTimeTimers();
+	}
+
+	public int numEventTimeTimers(W window) {
+		return internalTimerService.numEventTimeTimers(window);
+	}
+
+	public int numStateEntries() {
+		return stateBackend.numStateEntries();
+	}
+
+	public int numStateEntries(W window) {
+		return stateBackend.numStateEntries(window);
+	}
+
+	/**
+	 * Injects one element into the trigger for the given window and returns the result of
+	 * {@link Trigger#onElement(Object, long, Window, Trigger.TriggerContext)}
+	 */
+	public TriggerResult processElement(StreamRecord<T> element, W window) throws Exception {
+		TestTriggerContext<Integer, W> triggerContext = new TestTriggerContext<>(
+				KEY,
+				window,
+				internalTimerService,
+				stateBackend,
+				windowSerializer);
+		return trigger.onElement(element.getValue(), element.getTimestamp(), window, triggerContext);
+	}
+
+	/**
+	 * Advanced processing time and checks whether we have exactly one firing for the given
+	 * window. The result of {@link Trigger#onProcessingTime(long, Window, Trigger.TriggerContext)}
+	 * is returned for that firing.
+	 */
+	public TriggerResult advanceProcessingTime(long time, W window) throws Exception {
+		Collection<Tuple2<W, TriggerResult>> firings = advanceProcessingTime(time);
+
+		if (firings.size() != 1) {
+			throw new IllegalStateException("Must have exactly one timer firing. Fired timers: " + firings);
+		}
+
+		Tuple2<W, TriggerResult> firing = firings.iterator().next();
```

Comment:
With the check in the lines above (that checks for the size of `firings`) it should be correct, right?